woocommerce/woocommerce
# Product Collection: Trigger `wc-blocks_viewed_product` JS event

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR adds wc-blocks_viewed_product event being triggered from three places:

- Product Image
- Product Title
- Add to Cart when it's an anchor (e.g. "Select Options", "Read More", etc.)

This is the second iteration after https://github.com/woocommerce/woocommerce/pull/50217.

In opposite to the previous one, here we use a single action on the Product Collection store and trigger it from the inner blocks. Also, the Product Template provides a Interactivity context that keeps product ID.

Closes https://github.com/woocommerce/woocommerce/issues/48861



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites
1. Use some plugin to include custom JS script on the page e.g. [Easy Code Snippets](https://wordpress.org/plugins/easy-code-snippets/)
2. Attach this code:

```
<script>
document.body.addEventListener(
	'wc-blocks_viewed_product',
	e => console.log('Event triggered', e.detail)
);
</script>
```

If you're using Easy Code Snippets, don't forget to wrap it in `<script></script>` tag and set the position to "Footer":
<img width="1159" alt="image" src="https://github.com/user-attachments/assets/c6a726f0-b204-4770-aa70-c6499a92b0ec">

#### Testing
1. Make sure you have some variable, grouped and out of stock products
2. Create new post
3. Add Product Collection block
4. Choose some collection
4. Save and go to frontend
5. Open browser's console
6. Make sure you enable "Preserve log" option (example view from Chrome browser, but it may be in different place in your browser):

![Screen Shot 2024-08-28 at 10 15 31 AM](https://github.com/user-attachments/assets/63e9b7d5-701a-457b-90ad-a7bd4b729bec)

8. For variable, grouped, and out-of-stock products click their "Add to Cart button" which should have labels like
    - "Select options"
    - "View products"
    - "Read more"
9. **Expected:** Along with redirection there should be an event logged in your console. Event should look like this with the `productId` corresponding to the product you clicked and `collection` corresponding to collection you chose (it's `undefined` if you chose "create your own"):

<img width="738" alt="image" src="https://github.com/user-attachments/assets/d39b0837-7943-4433-9b06-5d120215df46">



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
